src/FileList/FileListHandler.cc

```diff
@@ -157,24 +178,58 @@ void FileListHandler::GetFileList(CARTA::FileListResponse& file_list, std::strin
                                 }
                             }
                         } else if (cc_file.isRegular(true) && cc_file.isReadable()) {
-                            if ((image_type == casacore::ImageOpener::FITS) || (image_type == casacore::ImageOpener::HDF5)) {
-                                add_image = true;
-                            } else if (region_list) { // list unknown files: name, type, size
-                                add_image = true;
+                            switch (image_type) {
+                                case casacore::ImageOpener::FITS: {
+                                    file_type = CARTA::FileType::FITS;
+                                    add_image = true;
+                                    break;
+                                }
+                                case casacore::ImageOpener::HDF5: {
+                                    file_type = CARTA::FileType::HDF5;
+                                    add_image = true;
+                                    break;
+                                }
+                                default: {
+                                    file_type = CARTA::FileType::UNKNOWN;
+                                    break;
+                                }
                             }
                         }
 
                         if (add_image) { // add image to file list
                             auto& file_info = *file_list.add_files();
                             file_info.set_name(name);
-                            FillFileInfo(file_info, full_path);
+                            FillFileInfo(file_info, full_path, file_type);
                         }
                     }
                 } catch (casacore::AipsError& err) { // RegularFileIO error
                     // skip it
                 }
             }
             dir_iter++;
+
+            ++num_of_files_done;
+            percentage = (float)num_of_files_done / (float)total_files;
+            auto current_time = std::chrono::high_resolution_clock::now();
+            auto dt = std::chrono::duration<double>(current_time - start_time).count();
+
+            auto report_progress = [&]() {
+                CARTA::ListProgress progress;
+                progress.set_file_list_type(CARTA::FileListType::Image);
+                progress.set_percentage(percentage);
+                progress.set_checked_count(num_of_files_done);
+                progress.set_total_count(total_files);
+                _progress_callback(progress);
+                start_time = current_time;
+            };
+
+            // report the progress if it fits the conditions
+            if (!_first_report && dt > REPORT_FIRST_PROGRESS_AFTER_SECS) {
```

Comment:
The name `_first_report` is misleading -- I suggest changing it to something like `_first_report_made`.

`REPORT_FIRST_PROGRESS_AFTER_SECS` is inconsistently named, and it's not clear that this has anything to do with the file list. There are also some unclearly named constants used for image moments (including one called `REPORT_FIRST_PROGRESS_AFTER_MILLI_SECS`!). But I think that the correct way to fix this is to refactor the constants file and move these into the scopes where they are actually used, which should happen in a separate PR.

Reply:
I see now that these are used in two different places, so the name should probably be fixed in the short term. I would change it to `FILE_LIST_FIRST_PROGRESS_AFTER_SECS`.

Reply:
I will refactor the constants for image moments in a separate PR.